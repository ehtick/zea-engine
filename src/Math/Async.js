import { Signal } from '../Math/Signal';

class Async {
    constructor() {
        this.__asyncCount = 0;
        this.ready = new Signal(true);

        this.incAsyncCount = function(count=1) {
            this.__asyncCount+=count;
        }.bind(this);

        this.decAsyncCount = function() {
            if (this.__asyncCount > 0){
                this.__asyncCount--;
                if (this.__asyncCount == 0) {
                    this.__asyncsCompleted();
                }
            }
        }.bind(this);

        this.__asyncsCompleted = function(){
            this.ready.emit();
        }.bind(this)
    }

    get count(){
        return this.__asyncCount;
    }

};

export {
    Async
};
// export default Async;

