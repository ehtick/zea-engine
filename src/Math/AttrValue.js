// eslint-disable-next-line camelcase
import { JSON_stringify_fixedPrecision } from './Common.js'

/**
 * Base class for values that can be stored in vertex attributes.
 * <br>
 * **Note:** These values use {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array|Float32Array} values to store their data.
 */
class AttrValue {
  /**
   * Verifies that all the numeric values inside the `this.__data` variable are a valid number.
   * Returns `false` If at least one of the values is either {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/Infinity|Infinity} or
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/NaN|NaN}.
   *
   * @return {boolean} - Returns the result as a boolean.
   */
  isValid() {
    for (const v of this.__data) {
      if (v == Infinity || isNaN(v)) return false
    }

    return true
  }

  /**
   * Creates a new value to wrap memory in an existing buffer.
   *
   * @param {ArrayBuffer} buffer - the buffer value.
   * @param {number} offset - the offset value.
   *
   * @todo Implement this function
   */
  static createFromFloat32Buffer(buffer, offset) {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns the number of Float32 elements used by this type. Used to calculate storage requirements for large arrays of this type.
   *
   * @todo Implement this function
   */
  static numElements() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns current Vec2 data as array. Often used to pass types to the GPU.
   *
   * @return {array} - Returns the result as an array.
   */
  asArray() {
    return this.__data
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString() {
    // eslint-disable-next-line new-cap
    return JSON_stringify_fixedPrecision(this.toJSON())
  }

  /**
   * Represents the state of your class as a JSON Object.
   *
   * @todo Implement this function in the derived class.
   */
  toJSON() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }
}

export { AttrValue }
export default AttrValue
