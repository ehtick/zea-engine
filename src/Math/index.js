
export {
    SInt32,
    UInt32,
    Float32,
    hashStr,
    JSON_stringify_fixedPrecision
} from './Common.js';
export * from './AttrValue.js';
export * from './Vec2.js';
export * from './Vec3.js';
export * from './Vec4.js';
export * from './Color.js';
export * from './Quat.js';
export * from './Ray.js';
export * from './Mat3.js';
export * from './Mat4.js';
export * from './Xfo.js';
export * from './Box2.js';
export * from './Box3.js';
export * from './Rect2.js';
export * from './Hammersley.js';

export * from './TypeRegistry.js';
