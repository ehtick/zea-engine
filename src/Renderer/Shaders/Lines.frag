
precision highp float;

import 'GLSLUtils.glsl'
import 'drawItemTexture.glsl'
import 'cutaways.glsl'
import 'materialparams.glsl'


uniform int occluded;

#ifndef ENABLE_MULTI_DRAW

uniform color BaseColor;
uniform float Opacity;

uniform float StippleScale;
uniform float StippleValue;
uniform float OccludedStippleValue;

#endif // ENABLE_MULTI_DRAW

#if defined(DRAW_GEOMDATA)

uniform int floatGeomBuffer;
uniform int passId;

import 'GLSLBits.glsl'

#elif defined(DRAW_HIGHLIGHT)

#ifdef ENABLE_FLOAT_TEXTURES
vec4 getHighlightColor(int id) {
  return fetchTexel(instancesTexture, instancesTextureSize, (id * pixelsPerItem) + 4);
}
#else // ENABLE_FLOAT_TEXTURES

uniform vec4 highlightColor;

vec4 getHighlightColor() {
  return highlightColor;
}

#endif // ENABLE_FLOAT_TEXTURES

#endif // DRAW_HIGHLIGHT


/* VS Outputs */
varying float v_geomItemId;
varying vec4 v_geomItemData;
varying vec3 v_viewPos;
varying vec3 v_worldPos;
varying vec3 v_nextVertexDist;

#ifdef ENABLE_ES3
  out vec4 fragColor;
#endif

void main(void) {
#ifndef ENABLE_ES3
  vec4 fragColor;
#endif

  int geomItemId = int(v_geomItemId + 0.5);
  int flags = int(v_geomItemData.r + 0.5);

  // Cutaways
  if (testFlag(flags, GEOMITEM_FLAG_CUTAWAY)) 
  {
    vec4 cutAwayData   = getCutaway(geomItemId);
    vec3 planeNormal = cutAwayData.xyz;
    float planeDist = cutAwayData.w;
    if (cutaway(v_worldPos, planeNormal, planeDist)) {
      discard;
      return;
    }
  }

  //////////////////////////////////////////////
  // Material

#ifdef ENABLE_MULTI_DRAW

  vec2 materialCoords = v_geomItemData.zw;
  vec4 BaseColor = getMaterialValue(materialCoords, 0);
  vec4 matValue1 = getMaterialValue(materialCoords, 1);
  vec4 matValue2 = getMaterialValue(materialCoords, 2);
  float Opacity  = matValue1.r;

  float StippleScale = matValue1.b;
  float StippleValue = matValue1.a;
  float OccludedStippleValue = matValue2.r;
#endif // ENABLE_MULTI_DRAW

  ///////////////////
  // Stippling
  float stippleValue = occluded == 0 ? StippleValue : OccludedStippleValue;
#ifdef ENABLE_ES3 // No stippling < es3 
  if (stippleValue > 0.0) {
    // Note: a value of 0.0, means no stippling (solid). A value of 1.0 means invisible
    float dist = -v_viewPos.z * StippleScale;
    float nextVertexDist = imod(int(floor(v_nextVertexDist.z)), 2) == 0 ? v_nextVertexDist.x : v_nextVertexDist.y;
    if (mod(nextVertexDist / dist, 1.0) < stippleValue) {
      discard;
      return;
    }
  }
#endif

  //////////////////////////////////////////////
  // Color
#if defined(DRAW_COLOR)

  fragColor = BaseColor;
  fragColor.a *= Opacity;

  
#ifndef ENABLE_ES3
  if (occluded == 1) fragColor.a *= 1.0 - stippleValue;
#endif

  //////////////////////////////////////////////
  // GeomData
#elif defined(DRAW_GEOMDATA)

  if (testFlag(flags, GEOMITEM_INVISIBLE_IN_GEOMDATA)) {
    discard;
    return;
  }
  
  float viewDist = length(v_viewPos);

  if (floatGeomBuffer != 0) {
    fragColor.r = float(passId); 
    fragColor.g = float(v_geomItemId);
    // Note: to make lines visually stand out from triangles
    // this value is 0.0 in the surface shaders.
    fragColor.b = 1.0;// TODO: store segment-id or something.
    fragColor.a = viewDist;
  } else {
    ///////////////////////////////////
    // UInt8 buffer
    fragColor.r = mod(v_geomItemId, 256.) / 256.;
    fragColor.g = (floor(v_geomItemId / 256.) + (float(passId) * 64.)) / 256.;

    // encode the dist as a 16 bit float
    vec2 float16bits = encode16BitFloatInto2xUInt8(viewDist);
    fragColor.b = float16bits.x;
    fragColor.a = float16bits.y;
  }

  //////////////////////////////////////////////
  // Highlight
#elif defined(DRAW_HIGHLIGHT)
  
  fragColor = getHighlightColor(geomItemId);

#endif // DRAW_HIGHLIGHT


#ifndef ENABLE_ES3
  gl_FragColor = fragColor;
#endif
}
