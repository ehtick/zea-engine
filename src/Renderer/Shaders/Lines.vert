
precision highp float;
precision highp int;

attribute vec3 positions;
attribute vec3 positionsNext;

import 'GLSLUtils.glsl'
import 'geomItemId.glsl'
import 'drawItemTexture.glsl'
import 'modelMatrix.glsl'

uniform mat4 viewMatrix;
uniform mat4 projectionMatrix;
uniform int isOrthographic;

#ifdef ENABLE_MULTI_DRAW
import 'materialparams.glsl'
#else
uniform float Overlay;
#endif

/* VS Outputs */
varying float v_geomItemId;
varying vec4 v_geomItemData;
varying vec3 v_viewPos;
varying vec3 v_worldPos;
varying vec3 v_nextVertexDist;

void main(void) {
  int geomItemId = getGeomItemId();
  v_geomItemId = float(geomItemId);
  v_geomItemData  = getInstanceData(geomItemId);

  mat4 modelMatrix = getModelMatrix(geomItemId);
  mat4 modelViewMatrix = viewMatrix * modelMatrix;
  vec4 viewPos = modelViewMatrix * vec4(positions, 1.0);
  vec4 viewPosNext = modelViewMatrix * vec4(positionsNext, 1.0);

#ifdef ENABLE_ES3
  float nextVertexDist = length(viewPosNext.xyz - viewPos.xyz);
  if (imod(gl_VertexID, 2) == 0) {
    v_nextVertexDist.x = nextVertexDist;
    v_nextVertexDist.y = 0.0;
  } else {
    v_nextVertexDist.x = 0.0;
    v_nextVertexDist.y = nextVertexDist;
  }
  v_nextVertexDist.z = float(gl_VertexID);
#endif

  v_viewPos = viewPos.xyz;
  gl_Position = projectionMatrix * viewPos;
    

  //////////////////////////////////////////////
  // Overlay

#ifdef ENABLE_MULTI_DRAW
  vec2 materialCoords = v_geomItemData.zw;
  vec4 materialValue1 = getMaterialValue(materialCoords, 1);
  float overlay = materialValue1.y;
#else
  float overlay = Overlay;
#endif

   
#if defined(DRAW_GEOMDATA)
  float _overlay = mix(overlay, 1.0, 0.0001);
#else
  float _overlay = overlay;
#endif

  if (isOrthographic > 0){
    gl_Position.z -= _overlay;
  } else {
    gl_Position.z = mix(gl_Position.z, -gl_Position.z, _overlay);
  }

  //////////////////////////////////////////////
  
  
  vec4 pos = vec4(positions, 1.);
  v_worldPos      = (modelMatrix * pos).xyz;
}
