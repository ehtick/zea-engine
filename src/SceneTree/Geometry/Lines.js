import { BaseGeom } from './BaseGeom.js';

class Lines extends BaseGeom {
    constructor(name) {
        super(name);
        this.__indices = new Uint32Array();
        this.__segmentAttributes = new Map();
        this.lineThickness = 0.0;
    }

    getIndices() {
        return this.__indices;
    }

    getNumSegments() {
        return this.__indices.length / 2;
    }

    setNumSegments(count) {
        let indices = new Uint32Array(count * 2);
        for (let i=0;i<this.__indices.length; i++) {
            indices[i] = this.__indices[i];
        }
        this.__indices = indices;
    }

    setSegment(index, p0, p1) {
        if (index >= (this.__indices.length / 2))
            throw ("Invalid line index:" + index + ". Num Segments:" + (this.__indices.length / 2));
        this.__indices[(index * 2) + 0] = p0;
        this.__indices[(index * 2) + 1] = p1;
    }

    getSegmentVertexIndex(line, linevertex) {
        let numLines = this.numLines;
        if (line < numLines)
            return this.__indices[(line * 2) + linevertex];
    }

    addSegmentAttribute(name, dataType, count = undefined) {
        let attr = new Attribute(dataType, (count != undefined) ? count : this.polygonCount);
        this.__segmentAttributes.set(name, attr);
        return attr;
    }

    hasSegmentAttribute(name) {
        return this.__segmentAttributes.has(name);
    }

    getSegmentAttribute(name) {
        return this.__segmentAttributes.get(name)
    }

    //////////////////////////////////////////
    // Memory
    
    genBuffers() {
        let buffers = super.genBuffers();

        let indices;
        if (buffers.numVertices < Math.pow(2, 8)){
            indices = new Uint8Array(this.__indices.length);
            this.__indices.forEach((value, index)=>{
                indices[index] = value;
            });
        }
        else if (totalNumVertices < Math.pow(2, 16)){
            indices = new Uint16Array(this.__indices.length);
            this.__indices.forEach((value, index)=>{
                indices[index] = value;
            });
        }
        else{
            indices = this.__indices;
        }
        buffers.indices = indices;
        return buffers;
    }

    //////////////////////////////////////////
    // Persistence

    toJSON() {
        let json = super.toJSON();
        json['indices'] = this.__indices.length;
        return json;
    };
};

export {
    Lines
};
//export default Lines;