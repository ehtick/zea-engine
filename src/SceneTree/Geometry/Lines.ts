/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { BaseGeom } from './BaseGeom'
import { Registry } from '../../Registry'
import { BinReader } from '../../SceneTree/BinReader'
import { Xfo } from '../../Math'
/**
 *
 * Class representing lines primitive drawing type, connecting vertices using the specified indices.
 * i.e. We have 4 points(vertices) but we don't know how they connect to each other,
 * and that's why we need indices(Numbers indicating which vertex connects to which).
 * In this case if we say that `indices` is `[0,1,2,3]`, it would connect the first vertex to the second,
 * and the third to the fourth.
 *
 * ```
 * const lines = new Lines()
 * ```
 *
 * **Events**
 * * **geomDataChanged:** Triggered when the data value of the geometry is set(This includes reading binary)
 *
 * @extends BaseGeom
 */
class Lines extends BaseGeom {
  protected __indices: Uint8Array | Uint16Array | Uint32Array

  /**
   * Create lines.
   */
  constructor() {
    super()
    this.__indices = new Uint32Array()
  }

  /**
   * The clear method.
   */
  clear(): void {
    this.setNumSegments(0)
    this.setNumVertices(0)

    this.emit('geomDataTopologyChanged')
  }

  /**
   * Returns the specified indices(Vertex connectors)
   *
   * @return - The indices index array.
   */
  getIndices(): Float32Array | Int32Array | Int16Array | Int8Array | Uint8Array | Uint16Array | Uint32Array {
    return this.__indices
  }

  /**
   * Returns the number of line segments.
   *
   * @return - Returns the number of segments.
   */
  getNumSegments(): number {
    return this.__indices.length / 2
  }

  /**
   * Returns the number of line segments.
   *
   * @return - Returns the number of segments.
   */
  getNumLineSegments(): number {
    return this.__indices.length / 2
  }

  /**
   * Sets the number of line segments in the lines geometry.
   * **Important:** It resets indices values.
   *
   * @param numOfSegments - The count value.
   */
  setNumSegments(numOfSegments: number): void {
    if (numOfSegments > this.getNumSegments()) {
      const indices = new Uint32Array(numOfSegments * 2)
      indices.set(this.__indices)
      this.__indices = indices
    } else {
      this.__indices = this.__indices.slice(0, numOfSegments * 2)
    }
  }

  /**
   * Sets segment values in the specified index.
   *
   * @param index - The index value.
   * @param p0 - The p0 value.
   * @param p1 - The p1 value.
   */
  setSegmentVertexIndices(index: number, p0: number, p1: number): void {
    if (index >= this.__indices.length / 2)
      throw new Error('Invalid line index:' + index + '. Num Segments:' + this.__indices.length / 2)
    this.__indices[index * 2 + 0] = p0
    this.__indices[index * 2 + 1] = p1
  }

  /**
   * The getSegmentVertexIndex method.
   *
   * @param line - The line value.
   * @param lineVertex - The lineVertex value.
   * @return - The return value.
   * @private
   */
  getSegmentVertexIndex(line: number, lineVertex: number): number {
    const numSegments = this.getNumSegments()
    if (line < numSegments) return this.__indices[line * 2 + lineVertex]
    return -1
  }

  /**
   * Merges a separate geometry into this one. Similar to a 'union' boolean operation.
   * @param other the other geom that will be merged into this one
   * @param xfo the transformation to be applied to the other geom as it is merged in.
   */
  merge(other: Lines, xfo: Xfo = new Xfo()) {
    const prevNumVerts = this.getNumVertices()
    super.merge(other, xfo)

    const otheIndices = other.__indices
    const indices = new Uint32Array(this.__indices.length + otheIndices.length)
    indices.set(this.__indices, 0)
    indices.set(
      otheIndices.map((index) => index + prevNumVerts),
      this.__indices.length
    )
    this.__indices = indices
  }

  // ////////////////////////////////////////
  // Memory

  /**
   * Returns vertex attributes buffers and its count.
   *
   * @return - The return value.
   */
  genBuffers(opts?: Record<string, any>): Record<string, any> {
    const buffers = super.genBuffers()

    let indices
    if (buffers.numVertices < Math.pow(2, 8)) {
      indices = new Uint8Array(this.__indices)
    } else if (buffers.numVertices < Math.pow(2, 16)) {
      indices = new Uint16Array(this.__indices)
    } else {
      indices = this.__indices
    }
    ;(buffers as any).indices = indices
    return buffers
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * Sets state of current geometry(Including line segments) using a binary reader object.
   *
   * @param reader - The reader value.
   * @param context - The context value.
   */
  readBinary(reader: BinReader, context?: Record<string, any>): void {
    super.loadBaseGeomBinary(reader, context)

    this.setNumSegments(reader.loadUInt32())

    const bytes = reader.loadUInt8()
    if (bytes == 1) this.__indices = reader.loadUInt8Array()
    else if (bytes == 2) this.__indices = reader.loadUInt16Array()
    else if (bytes == 4) this.__indices = reader.loadUInt32Array()

    this.emit('geomDataChanged')
  }
  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param context - The context value.
   * @return - Returns the json object.
   */
  toJSON(context?: Record<string, any>): Record<string, any> {
    const j = super.toJSON(context)
    if (!context || !context.skipTopology) (j as any).indices = Array.from(this.__indices)
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param j - The json object this item must decode.
   * @param context - The context value.
   */
  fromJSON(j: Record<string, any>, context?: Record<string, any>): void {
    super.fromJSON(j, context)
    if (j.indices) this.__indices = Uint32Array.from(j.indices)
  }
}

Registry.register('Lines', Lines)

export { Lines }
