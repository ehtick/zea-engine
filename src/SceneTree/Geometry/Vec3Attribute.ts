import { Attribute } from './Attribute'
import { Vec3 } from '../../Math/Vec3'
import { Registry } from '../../Registry'

/**
 * Class representing an attribute.
 */
class Vec3Attribute extends Attribute {
  /**
   * Create a Vec3Attribute.
   */
  constructor() {
    super('Vec3', 3)
    this.normalized = false
  }

  /**
   * Returns the Vec3 value at the specified index.
   *
   * > Note: 'Ref' means that the value contains a reference to the data in the attribute.
   * > The components of the value can be changed causing the attributes data is changed.
   * > No need to call 'setValue'.
   *
   * @param {number} index - The index value.
   * @returns Vec3 - The value at the specified index.
   */
  getValueRef(index: number): Vec3 {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    const valueData = this.data.subarray(offset, offset + this.stride)
    return new Vec3(valueData)
  }

  /**
   * Returns a copy of the Vec3 value at the specified index.
   *
   * @param {number} index - The index value.
   * @return Vec3 - The value at the specified index.
   */
  getValue(index: number): Vec3 {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    const valueData = this.data.slice(offset, offset + this.stride)
    return new Vec3(valueData)
  }

  /**
   * Sets Vec3 at the specified index.
   *
   * @param {number} index - The index value.
   * @param {Vec3} value - The value param.
   */
  setValue(index: number, value: Vec3): void {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    this.data.set(value.asArray(), offset)
  }

  /**
   * Gets the value of a corner vertex of a face.
   * > Note: 'Ref' means that the value contains a reference to the data in the attribute.
   * > The components of the value can be changed causing the attributes data is changed.
   * > No need to call 'setFaceVertexValue'.
   * @param {number} face - The face index.
   * @param {number} faceVertex - The index of vertex within the face. [0... num face vertices]
   * @return {Vec3} - The return value.
   */
  getFaceVertexValueRef(face: number, faceVertex: number): any {
    const array = this.getFaceVertexValueRef_array(face, faceVertex)
    return new Vec3(array)
  }

  /**
   * Sets the value of a corner vertex of a face.
   * @param {number} face - The face index.
   * @param {number} faceVertex - The index of vertex within the face. [0... num face vertices]
   * @param {Vec3} value - The value value.
   */
  setFaceVertexValue(face: number, faceVertex: number, value: Vec3): void {
    this.setFaceVertexValue_array(face, faceVertex, value.asArray())
  }

  /**
   * The setSplitVertexValue method.
   * @param {number} vertex - The vertex value.
   * @param {number} face - The face index.
   * @param {any} value - The value value.
   */
  setSplitVertexValue(vertex: number, face: number, value: Vec3): void {
    this.setSplitVertexValue_array(vertex, face, value.asArray())
  }
}

export { Vec3Attribute }
