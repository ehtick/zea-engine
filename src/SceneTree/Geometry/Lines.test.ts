import { Lines } from './Lines'
import { Vec3 } from '../../Math'
import { Vec3Attribute } from './Vec3Attribute'

describe('Lines', () => {
  test('Check for default positions attribute.', () => {
    const lines = new Lines()
    expect(lines.hasVertexAttribute('positions')).toBe(true)
  })

  test('Check for positions attribute has the same length as the number of vertices.', () => {
    const lines = new Lines()
    const numVertices = 3
    lines.setNumVertices(numVertices)

    const positions = <Vec3Attribute>lines.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    lines.setSides(2)
    lines.setSegmentVertexIndices(0, 0, 1)
    lines.setSegmentVertexIndices(1, 1, 2)

    expect(lines.getSegmentVertexIndex(1, 0)).toBe(1)
    expect(lines.getSegmentVertexIndex(1, 1)).toBe(2)

    expect(lines.getVertexAttribute('positions').getCount()).toBe(numVertices)
  })

  test('Check resizing bigger the line segment indices.', () => {
    const lines = new Lines()
    const numVertices = 3
    lines.setNumVertices(numVertices)

    lines.setSides(2)
    lines.setSegmentVertexIndices(0, 0, 1)
    lines.setSegmentVertexIndices(1, 1, 2)

    lines.setSides(3)
    lines.setSegmentVertexIndices(2, 2, 0)

    expect(lines.getSegmentVertexIndex(1, 0)).toBe(1)
    expect(lines.getSegmentVertexIndex(1, 1)).toBe(2)

    // TODO:  originally: expect(lines.getIndices(1,1).toEqual(new Uint32Array([0, 1, 1, 2, 2, 0]))
    let indices = lines.getIndices()
    let expected = new Uint32Array([0, 1, 1, 2, 2, 0])
    expect(indices.toString()).toEqual(expected.toString())
  })

  test('Check resizing smaller the line segment indices.', () => {
    const lines = new Lines()
    const numVertices = 3
    lines.setNumVertices(numVertices)

    lines.setSides(3)
    lines.setSegmentVertexIndices(0, 0, 1)
    lines.setSegmentVertexIndices(1, 1, 2)
    lines.setSegmentVertexIndices(2, 2, 0)

    lines.setSides(2)

    expect(lines.getSegmentVertexIndex(1, 0)).toBe(1)
    expect(lines.getSegmentVertexIndex(1, 1)).toBe(2)
    expect(lines.getIndices().toString()).toEqual(new Uint32Array([0, 1, 1, 2]).toString())
  })

  test('Check generated buffers', () => {
    const lines = new Lines()
    const numVertices = 3
    lines.setNumVertices(numVertices)

    const positions = <Vec3Attribute>lines.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    lines.setSides(2)
    lines.setSegmentVertexIndices(0, 0, 1)
    lines.setSegmentVertexIndices(1, 1, 2)

    expect(lines.getSegmentVertexIndex(1, 0)).toBe(1)
    expect(lines.getSegmentVertexIndex(1, 1)).toBe(2)

    expect(lines.genBuffers()).toEqual({
      attrBuffers: {
        positions: {
          count: 3,
          dataType: 'Vec3',
          normalized: false,
          values: new Float32Array([1, 2, 3, -1, -2, -3, 2, 1, -3]),
        },
      },
      indices: new Uint8Array([0, 1, 1, 2]),
      numVertices: 3,
    })
  })

  it('saves to JSON (serialization).', () => {
    const lines = new Lines()
    const numVertices = 3
    lines.setNumVertices(numVertices)
    const positions = <Vec3Attribute>lines.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    lines.setSides(2)
    lines.setSegmentVertexIndices(0, 0, 1)
    lines.setSegmentVertexIndices(1, 1, 2)

    expect(JSON.stringify(lines.toJSON())).toMatchSnapshot()
  })

  it('loads from JSON (serialization).', () => {
    const lines = new Lines()
    const input = {
      type: 'Lines',
      numVertices: 3,
      vertexAttributes: {
        positions: { data: [1, 2, 3, -1, -2, -3, 2, 1, -3], dataType: 'Vec3', defaultValue: 0, length: 3 },
      },
      indices: [0, 1, 1, 2],
    }
    lines.fromJSON(input)

    expect(lines.getVertexAttribute('positions').getCount()).toBe(3)
    expect(lines.getNumSegments()).toBe(2)
  })
})
