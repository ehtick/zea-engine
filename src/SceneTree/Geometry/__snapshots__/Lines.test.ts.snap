// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lines saves to JSON (serialization). 1`] = `"{\\"type\\":\\"Lines\\",\\"numVertices\\":3,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":{\\"0\\":1,\\"1\\":2,\\"2\\":3,\\"3\\":-1,\\"4\\":-2,\\"5\\":-3,\\"6\\":2,\\"7\\":1,\\"8\\":-3},\\"dataType\\":\\"Vec3\\",\\"length\\":3}},\\"indices\\":[0,1,1,2]}"`;
